Firmware/HomeButtonsArduino/src/app.cpp

```diff
@@ -933,6 +555,423 @@ void App::_net_on_connect() {
 
   if (device_state_.persisted().send_discovery_config) {
     device_state_.persisted().send_discovery_config = false;
+    info("Sending discovery config...");
     mqtt_.send_discovery_config();
   }
 }
+
+void App::_download_mdi_icons() {
+  bool download_required = false;
+  mdi_.begin();
+  for (uint8_t i = 0; i < NUM_BUTTONS; i++) {
+    ButtonLabel label(device_state_.get_btn_label(i).c_str());
+    if (label.substring(0, 4) == "mdi:") {
+      MDIName icon = label.substring(
+          4, label.index_of(' ') > 0 ? label.index_of(' ') : label.length());
+      if (!mdi_.exists(icon.c_str())) {
+        download_required = true;
+        break;
+      }
+    }
+  }
+  if (!download_required) {
+    info("no icons to download");
+    mdi_.end();
+    return;
+  }
+
+  display_.disp_message("Downloading\nicons...");
+
+  // check if server is reachable
+  if (mdi_.check_connection()) {
+    info("icon server reachable");
+  } else {
+    warning("icon server NOT reachable");
+    mdi_.end();
+    delay(100);
+    display_.disp_error("Icon\nserver\nNOT\nreachable");
+    device_state_.flags().display_redraw = true;
+    return;
+  }
+
+  // free up space if needed
+  size_t free = mdi_.get_free_space();
+  info("SPIFFS free space: %d", free);
+  if (free < 100000UL) {
+    info("making space...");
+    if (!mdi_.make_space(100000UL)) {
+      error("failed to make space");
+      mdi_.end();
+      return;
+    }
+  }
+
+  info("Downloading icons...");
+  for (uint8_t i = 0; i < NUM_BUTTONS; i++) {
+    ButtonLabel label(device_state_.get_btn_label(i).c_str());
+    if (label.substring(0, 4) == "mdi:") {
+      MDIName icon = label.substring(
+          4, label.index_of(' ') > 0 ? label.index_of(' ') : label.length());
+      if (!mdi_.exists(icon.c_str())) {
+        mdi_.download(icon.c_str());
+      }
+    }
+  }
+  mdi_.end();
+  device_state_.flags().display_redraw = true;
+}
+
+void AppSMStates::InitState::entry() {
+  sm().network_.connect();
+
+  sm()._start_button_task();
+  sm()._start_leds_task();
+  sm()._start_display_task();
+  sm()._start_network_task();
+
+  sm().mdi_.add_size(64);
+  sm().mdi_.add_size(48);
+
+  if (!sm().device_state_.flags().awake_mode) {
+    esp_task_wdt_init(WDT_TIMEOUT_SLEEP, true);
+    esp_task_wdt_add(NULL);
+    if (sm().boot_cause_ == BootCause::TIMER) {
+      transition_to<NetConnectingState>();
```

Comment:
Please add `return` before all `transition_to`.
Even if it does not return anything (`void`), it will avoid executing code that was not intended to be.
I see at least one example (`AppSMStates::AwakeModeIdleState::loop()`) where multiple `transition_to` can be reached in one loop.